pkgs/tools/virtualization/nixos-nspawn/nixos-nspawn.py

```diff
@@ -0,0 +1,512 @@
+[Network]
+{addr_text}
+DHCPServer=yes
+EmitLLDP=customer-bridge
+IPForward=yes
+{nat_text}
+LLDP=yes
+""")
+
+            run(["systemctl", "restart", "systemd-networkd"])
+
+        with open(self.path, 'w+') as f:
+            f.write(f"""
+[Exec]
+Boot=false
+Parameters={path_}/init
+PrivateUsers=yes
+X-ActivationStrategy={data['activation']['strategy']}
+X-Imperative=1
+{ephemeral}
+{link_journal}
+[Files]
+BindReadOnly=/nix/store
+BindReadOnly=/nix/var/nix/db
+BindReadOnly=/nix/var/nix/daemon-socket
+PrivateUsersChown=yes
+[Network]
+{zone}
+{port_text}
+{private_network_text}
+""")
+
+        logging.info(f"Creating state-directory {self.name}")
+        makedirs(f"/var/lib/machines/{self.name}/etc", exist_ok=True)
+        if not update:
+            open(f"/var/lib/machines/{self.name}/etc/os-release", 'w').close()
+
+        self.pending = False
+        sync()
+
+        if update:
+            self.__activate(data, path_, strategy)
+        else:
+            logging.info(f"Booting new container {self.name}")
+            run(["machinectl", "start", self.name])
+
+    def __activate(self, data, path, strategy):
+        if strategy is None:
+            strategy = data['activation']['strategy']
+        if strategy == 'restart':
+            logging.info(f"Rebooting container {self.name}")
+            run(["machinectl", "reboot", self.name])
+        else:
+            logging.info(f"Reloading container {self.name}")
+            pid, _ = run(
+                ['machinectl', 'show', self.name, '--property', 'Leader', '--value'],
+                return_stdout=True
+            )
+            run([
+                "nsenter",
+                "-t",
+                pid.strip(),
+                "-m",
+                "-u",
+                "-U",
+                "-i",
+                "-n",
+                "-p",
+                "--",
+                f"{path}/bin/switch-to-configuration",
+                "test"
+            ])
+
+    def __machinectl_property(self, prop, default='<unknown>'):
+        str_ret, exit = run(
+            ['machinectl', 'show', self.name, '--property', prop, '--value'],
+            must_succeed=False,
+            return_stdout=True
+        )
+
+        if exit != 0:
+            return default
+
+        return str_ret
+
+    def __build_config(self):
+        def extract(key, config):
+            if key in config:
+                return config[key]
+            return {}
+
+        parser = MyConfigParser(
+            dict_type=SystemdSettings,
+            strict=False,
+            interpolation=Interpolation()
+        )
+        parser.read(self.path)
+        parser.sections()
+
+        self.exec = extract('Exec', parser)
+        self.files = extract('Files', parser)
+        self.network = extract('Network', parser)
+
+    def __profile_dir(self):
+        return f"{PROFILE_BASE}/{self.name}"
+
+    def __nix_path(self):
+        return path.realpath(f"{self.__profile_dir()}/system")
```

Comment:
Would it not be useful to keep these as public functions, or even better public properties?